### PR TITLE
Creating context from raw pointer

### PR DESCRIPTION
**Pull request by elichai**
Hi,
like @stevenroose asked in #178 
1. Would like feedback on the way it's implemented (*not* letting the user choose the exact generic, and not introducing a new trait).
2. I still have some open questions:
    - Do we want to guarantee what variants/invariants must hold here? (it seems like saying the list isn't exhaustive is somewhat not-useful)
    - Is there a way to check the correct capabilities in upstream and return a `Result`?
    - When and If we'll update to 1.25+ we should at least replace this with `NonNull`.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -751,6 +751,41 @@ mod tests {
...
+        // Try signing
+        assert_eq!(sign.sign(&msg, &sk), full.sign(&msg, &sk));
+        let sig = full.sign(&msg, &sk);
+
+        // Try verifying
+        assert!(vrfy.verify(&msg, &sig, &pk).is_ok());
+        assert!(full.verify(&msg, &sig, &pk).is_ok());
+
+        drop(full);drop(sign);drop(vrfy);
+        drop(ctx_full);drop(ctx_sign);drop(ctx_vrfy);
```
Comment (elichai) [inline src/lib.rs]: Probably not needed. But the manual drops here are meant to try and make sure the `ctx_*` don't drop before the Context's using it.

**Comment by apoelstra** [comment]
This looks good to me.

I think the list of invariants actually is exhaustive, for what it's worth.

**Comment by elichai** [comment]
Yeah, I wanted to be cautious with words on purpose, and then let other people way it on what exactly are we promising and what not.

**Comment by apoelstra** [comment]
Can you remove the "not exhaustive" bullet point?

**Comment by apoelstra** [comment]
So, after actually trying to use this, the semantics are not quite right.

We say "The user must handle the freeing of the context(using the correct functions) by himself" but in the `Drop` code for the returned object, we [still call `secp256k1_context_preallocated_destroy](https://github.com/elichai/rust-secp256k1/blob/9523e093c1c9a331fb120a101ef729a611a1ce55/src/lib.rs#L574). This means that if you get a raw pointer to a secp context you don't own, then call `from_raw_all` to get a `Secp256k1` object, you need to use `mem::forget` to avoid destroying the underlying context object.

**Comment by elichai** [comment]
Hmmm. That's right.
I wonder what's the right way here, we could check for nulness of the data buffer in the drop impl, Altough even though that's the case right now it might not be in the future.
We could add an `owner` bool.
Or we could add yet another trait.

**Comment by apoelstra** [comment]
I think checking for nulness is wrong and likely to cause problems.

Probably we should just add a boolean. This is a bit wasteful since we could do it with zero overhead if we had another trait, but our type system here is already pretty messy.

**Comment by elichai** [comment]
We could also add another struct instead. `ManualContext` that Derefs to Secp256k1 but contains it as a ManualDrop<Secp256k1> and return that instead of the regular context

**Comment by apoelstra** [comment]
Hm, we could also advise the user to just wrap the return value of the existing functions in `ManuallyDrop`.

**Comment by real-or-random** [comment]
> Hm, we could also advise the user to just wrap the return value of the existing functions in `ManuallyDrop`.

Could we instead do this for the user in `from_raw_all` etc? This will change the API types of course.

**Comment by stevenroose** [comment]
"Advise the user" is what often goes wrong :p But well in this case for a very niche use-case, it could be enough.

**Comment by stevenroose** [comment]
> Could we instead do this for the user in from_raw_all etc? This will change the API types of course.

Heh yeah `from_raw_all` could just be `-> ManuallyDrop<Secp256k1>`. Not too pretty of an API, but it'd do. Then the occasional user that _wants_ it to be automatically dropped, can call `.into_inner()`.

**Comment by apoelstra** [comment]
Yeah I think this is a good idea.

**Comment by apoelstra** [comment]
to change the API

**Pull request merged**